clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -117,15 +117,21 @@ public ByteBuffer allocate(int size, long maxTimeToBlock) throws InterruptedExce
                 int accumulated = 0;
                 ByteBuffer buffer = null;
                 Condition moreMemory = this.lock.newCondition();
+                long remainingTimeToBlockNs = TimeUnit.MILLISECONDS.toNanos(maxTimeToBlockMs);
                 this.waiters.addLast(moreMemory);
                 // loop over and over until we have a buffer or have reserved
                 // enough memory to allocate one
                 while (accumulated < size) {
-                    long startWait = time.nanoseconds();
-                    if (!moreMemory.await(maxTimeToBlock, TimeUnit.MILLISECONDS))
-                        throw new TimeoutException("Failed to allocate memory within the configured max blocking time");
-                    long endWait = time.nanoseconds();
-                    this.waitTime.record(endWait - startWait, time.milliseconds());
+                    long startWaitNs = time.nanoseconds();
+                    boolean waitingTimeElapsed = !moreMemory.await(remainingTimeToBlockNs, TimeUnit.NANOSECONDS);
+                    long endWaitNs = time.nanoseconds();
+                    long timeNs = Math.max(0L, endWaitNs - startWaitNs);
+                    this.waitTime.record(timeNs, TimeUnit.NANOSECONDS.toMillis(endWaitNs));
```

Comment:
Yes, that's what I mean. The issue is not the accuracy, but that `System.nanoTime` is not specified to return wall-clock time. So, unfortunately we need `time.milliseconds` (which is the same as `System.currentTimeMillis`). My original suggestion had the goal of saving the system call indeed, but that's not possible while maintaining the current semantics. Unfortunately, our `Time` interface is misleading in this respect (I filed https://issues.apache.org/jira/browse/KAFKA-2607 a while back to improve this).